### PR TITLE
RPN loss

### PR DESCRIPTION
**Pull request by 0x00b1**

### CODE REVIEW

File: keras_rcnn/losses/rpn.py [hunk 1]
```diff
@@ -1,21 +1,107 @@
 import keras
+import keras_rcnn.backend
+import numpy
+import tensorflow
 
+def separate_pred(y_pred):
+    n_anchors = tensorflow.shape(y_pred)[-1] // 5
```
Comment (0x00b1) [inline keras_rcnn/losses/rpn.py]:
```Python
n_anchors = keras.backend.shape(y_pred)[-1] // 5
```

File: keras_rcnn/losses/rpn.py [hunk 2]
```diff
@@ -1,21 +1,107 @@
...
-#     return f
+
+def encode(features, image_shape, y_true, stride = 16):
+    # get all anchors inside bbox
+    shifted_anchors = keras_rcnn.backend.shift(features, stride)
+    inds_inside, all_inside_bbox = keras_rcnn.backend.inside_image(shifted_anchors, image_shape)
+
+    # indices of gt boxes with the greatest overlap, bbox labels
+    # TODO: assert y_true.shape[0] == 1
+    y_true = tensorflow.gather(y_true, tensorflow.constant(0))
```
Comment (0x00b1) [inline keras_rcnn/losses/rpn.py]:
```Python
y_true = keras.backend.gather(y_true, keras.backend.constant(0))
```

File: keras_rcnn/losses/rpn.py [hunk 3]
```diff
@@ -1,21 +1,107 @@
...
+    argmax_overlaps_inds, bbox_labels = keras_rcnn.backend.label(y_true, all_inside_bbox, inds_inside)
+
+    # gt boxes
+    gt_boxes = tensorflow.gather(y_true, argmax_overlaps_inds)
```
Comment (0x00b1) [inline keras_rcnn/losses/rpn.py]:
```Python
gt_boxes = keras.backend.gather(y_true, argmax_overlaps_inds)
```

File: keras_rcnn/losses/rpn.py [hunk 4]
```diff
@@ -1,21 +1,107 @@
...
+
+    # Convert fixed anchors in (x, y, w, h) to (dx, dy, dw, dh)
+    bbox_reg_targets = keras_rcnn.backend.bbox_transform(all_inside_bbox, gt_boxes)
+
+    return bbox_labels, bbox_reg_targets, inds_inside, tensorflow.shape(shifted_anchors)[0]
```
Comment (0x00b1) [inline keras_rcnn/losses/rpn.py]:
```Python
return bbox_labels, bbox_reg_targets, inds_inside, keras.backend.shape(shifted_anchors)[0]
```

File: keras_rcnn/losses/rpn.py [hunk 5]
```diff
@@ -1,21 +1,107 @@
...
+        # separate y_pred into rpn_cls_pred and rpn_reg_pred
+        y_pred_regression, y_pred_classification = separate_pred(y_pred)
+
+        # convert y_true from gt_boxes to gt_anchors
+        if 'image_shape' in kwargs:
+            image_shape = kwargs['image_shape']
+        if 'stride' in kwargs:
+            stride = kwargs['stride']
+
+        features = tensorflow.shape(y_pred)[1:3]
```
Comment (0x00b1) [inline keras_rcnn/losses/rpn.py]:
```Python
features = keras.backend.shape(y_pred)[1:3]
```

File: keras_rcnn/losses/rpn.py [hunk 6]
```diff
@@ -1,21 +1,107 @@
...
+
+        gt_classification, gt_regression, inds_inside, num_shifted_anchors = encode(features, image_shape, y_true, stride)
+
+        N = features[0] * features[1] * anchors
+        ii = tensorflow.constant(0)
```
Comment (0x00b1) [inline keras_rcnn/losses/rpn.py]:
```Python
ii = keras.backend.constant(0)
```
